### PR TITLE
[WIP] Support to auto-generate Java files from flatbuffer

### PR DESCRIPTION
**Pull request by jovany-wang**
## What do these changes do?
1. generate java files from flatbuffer automatically.
2. add `auto_gen_tool.py` script to change files automatically.

After this PR, We should not change them manually.

### CODE REVIEW

File: src/ray/gcs/CMakeLists.txt [hunk 1]
```diff
@@ -30,7 +30,18 @@ add_custom_command(
   TARGET gen_gcs_fbs
   COMMAND ${FLATBUFFERS_COMPILER} -p -o ${PYTHON_OUTPUT_DIR} ${GCS_FBS_SRC}
   DEPENDS ${FBS_DEPENDS}
-  COMMENT "Running flatc compiler on ${GCS_FBS_SRC}"
+  COMMENT "Running flatc compiler on ${GCS_FBS_SRC} for python"
+  VERBATIM)
+
+# Generate Java bindings for the flatbuffers objects.
+set(RAY_HOME ${CMAKE_CURRENT_LIST_DIR}/../../..)
```
Comment (guoyuhong) [inline src/ray/gcs/CMakeLists.txt]: If you add `RAY_HOME`, you can also apply it to `set(PYTHON_OUTPUT_DIR ${CMAKE_CURRENT_LIST_DIR}/../../../python/ray/core/generated/)`.

File: auto_gen_tool.py [hunk 2]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
```
Comment (guoyuhong) [inline auto_gen_tool.py]: This comment is not necessary?

File: auto_gen_tool.py [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                continue
+            elif len(results) == 1:
+                return int(results[0])
+            else:
+                return -1
+
+        return -1
+
+
+template = '''
```
Comment (guoyuhong) [inline auto_gen_tool.py]: `template` looks too general here. Shall we use `template_for_byte_buffer_getter`?

**Review by guoyuhong — COMMENTED** [review]
Thanks. I left some comments.

File: auto_gen_tool.py [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+            file_handler.write(line)
+
+    return True
+
+
+def add_package_declarations(generated_root_path):
+    file_names = os.listdir(generated_root_path)
+    for file_name in file_names:
+        if not file_name.endswith('.java'):
+            break
```
Comment (raulchen) [inline auto_gen_tool.py]: this should be `continue`?

File: auto_gen_tool.py [hunk 5]
```diff
@@ -0,0 +1,89 @@
+#!/usr/bin/env python
+import os
+import re
+import sys
+
+
+def add_new_line(file, line_num, text):
+    with open(file, "r") as file_handler:
+        lines = file_handler.readlines()
+        if (line_num < 0) or (line_num > len(lines) + 1):
```
Comment (raulchen) [inline auto_gen_tool.py]: `<= 0`

File: auto_gen_tool.py [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def generate_and_insert_method(file, field, method_name):
+    index_to_be_inserted = -1
+    with open(file, 'r') as file_handler:
+        lines = file_handler.readlines()
+
+        for index in range(len(lines) - 1, -1, -1):
+            if lines[index] == '}\n':
+                index_to_be_inserted = index
```
Comment (raulchen) [inline auto_gen_tool.py]: why can't `index_to_be_inserted` just be `len(lines) - 1`?

File: auto_gen_tool.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+    if index_to_be_inserted >= 0:
+        offset = get_offset(file, field)
```
Comment (raulchen) [inline auto_gen_tool.py]: we should raise an exception if `get_offset` fails to find the line number

File: auto_gen_tool.py [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+if __name__ == '__main__':
+    ray_home = sys.argv[1]
+    root_path = '%s/java/runtime/src/main/java/org/ray/runtime/generated' % ray_home
+
+    add_package_declarations(root_path)
+    generate_and_insert_method('%s/TaskInfo.java' % root_path,
+                               'returns', 'returnsAsByteBuffer')
+    generate_and_insert_method('%s/Arg.java' % root_path,
+                               'objectIds', 'objectIdAsByteBuffer')
```
Comment (raulchen) [inline auto_gen_tool.py]: Let's define these fields as a list. E.g.,
```python
files_and_fields = [
   ('TaskInfo', 'returns'),
   ('Arg', 'objectIds'),
]
```

File: src/ray/gcs/CMakeLists.txt [hunk 9]
```diff
@@ -30,7 +30,18 @@ add_custom_command(
...
+set(JAVA_OUTPUT_DIR ${RAY_HOME}/java/runtime/src/main/java/org/ray/runtime/generated/)
+add_custom_command(
+  TARGET gen_gcs_fbs
+  COMMAND ${FLATBUFFERS_COMPILER} -j -o ${JAVA_OUTPUT_DIR} ${GCS_FBS_SRC}
+  COMMAND python ${RAY_HOME}/auto_gen_tool.py ${RAY_HOME}
```
Comment (raulchen) [inline src/ray/gcs/CMakeLists.txt]: Let's put the script under `java/` dir, and name it `modify_generated_java_flatbuffers_files.py`

**Review by raulchen — CHANGES_REQUESTED** [review]

File: auto_gen_tool.py (hunk 6 above)
Comment (jovany-wang) [reply auto_gen_tool.py]: Because we can not make sure that these is no blank line in the end of file. Maybe it have 2 blank.

**Comment by jovany-wang** [comment]
Thanks for you comments, @guoyuhong @raulchen 
I have addressed them. :)

**Comment by suquark** [comment]
I have added this PR to the project board since it removes magic flatbuffer files. We could probably mention it in the java docs for later developers. In the future we could ask the flatbuffer team to solve this problem.

**Comment by suquark** [comment]
And I am agree with @raulchen that we should rename `auto_gen_tool.py` to `modify_generated_java_flatbuffers_files ` and put it under the `java` dir.

**Comment by jovany-wang** [comment]
> And I am agree with @raulchen that we should rename `auto_gen_tool.py` to `modify_generated_java_flatbuffers_files ` and put it under the `java` dir.

Thanks for your mentioning again. I have addressed it. :)

**Comment by robertnishihara** [comment]
@jovany-wang @raulchen given the discussion in https://github.com/google/flatbuffers/issues/5092, isn't the simplest solution to define a separate table in order to implement a 2D array of bytes? E.g.,

```
table Bar {
   data: [ubyte];
}
table MyTable {
   bar: [Bar];
}
```

as @raulchen suggested. That would be fairly unintrusive, right? Then we don't need to modify the autogenerated files at all. That seems like a more sustainable approach.

Also, which flatbuffer schema is the one that actually requires the 2D array of bytes?

**Review by robertnishihara — COMMENTED** [review]
Ok, this is much better than than the current state of affairs, so I think we should do this, but longer term we really shouldn't modify the autogenerated files even if it makes the schema a little bit uglier. 

E.g., another option is to have a single string which is just all of the object IDs concatenated together.

A heavier weight option is to switch to protobuf.

File: src/ray/gcs/CMakeLists.txt [hunk 10]
```diff
@@ -23,14 +23,34 @@ add_custom_command(
...
-  COMMENT "Running flatc compiler on ${GCS_FBS_SRC}"
+  COMMENT "Running flatc compiler on ${GCS_FBS_SRC} for python"
+  VERBATIM)
+
+# Generate Java bindings for the flatbuffers objects.
+# This command will modify the auto-generated Java flatbuffer files for 2 reasons:
+#
+# 1) We have 2D array fields in Table TaskInfo and Arg in `gcs.fbs`, and we
+#    have no better approach to get the element from 2D array field in flatbuffer.
+#    See https://github.com/google/flatbuffers/issues/5092 for more details.
```
Comment (raulchen) [inline src/ray/gcs/CMakeLists.txt]:
```
In `gcs.fbs`, some fields are defined as `[string]`. And currently flatbuffers doesn't have an API to
get a single element in the list as raw bytes. See https://github.com/google/flatbuffers/issues/5092
for more details.
```

File: src/ray/gcs/CMakeLists.txt [hunk 11]
```diff
@@ -23,14 +23,34 @@ add_custom_command(
...
+#
+# 2) The package declaration in Java is different from python and C++, and there is
+#    no approach to specify package(namepsace) for Java specially.
```
Comment (raulchen) [inline src/ray/gcs/CMakeLists.txt]: `no approach` -> `no option in the flatc command`

File: java/modify_generated_java_flatbuffers_files.py [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+                continue
+            elif len(results) == 1:
+                return int(results[0])
+            else:
+                return -1
+
+        return -1
+
+
+template_for_byte_buffer_getter = """
```
Comment (raulchen) [inline java/modify_generated_java_flatbuffers_files.py]: upper-case the var name

File: java/modify_generated_java_flatbuffers_files.py [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+    ByteBuffer src = bb.duplicate().order(ByteOrder.LITTLE_ENDIAN);
+    int length = src.getInt(offset);
+    src.position(offset + 4);
+    src.limit(offset + 4 + length);
+    return src;
+  }
+"""
+
+
+def generate_and_insert_method(file, field, method_name):
```
Comment (raulchen) [inline java/modify_generated_java_flatbuffers_files.py]: genrate_byte_buffer_getter

**Comment by robertnishihara** [comment]
@jovany-wang @raulchen what do you think about replacing `new_actor_handles: [string];` with `new_actor_handles: string;`, where the string is simply the concatenation of all of the relevant strings? Then we wouldn't need to modify the autogenerated files.

This doesn't address point 2, but it addresses point 1 and we can probably find a way to address point 2.

**Comment by raulchen** [comment]
> @jovany-wang @raulchen what do you think about replacing `new_actor_handles: [string];` with `new_actor_handles: string;`, where the string is simply the concatenation of all of the relevant strings? Then we wouldn't need to modify the autogenerated files.
> 
> This doesn't address point 2, but it addresses point 1 and we can probably find a way to address point 2.

I think that' fine as well. (probably it's better to use `[ubytes]`, if that doesn't require too many code changes.)

For point 2, I think we still need a patch for now. But this patch is more stable, because it only inserts a line at the top, and doesn't depend on the content of the generated files.

**Comment by jovany-wang** [comment]
@robertnishihara @raulchen 
Thanks for your suggestions, I will work for this issue.

**Comment by pcmoritz** [comment]
I'm in favor of either defining

```
table ObjectID {
id: [ubyte];
}
```
(in the C++ codebase I think it could be done with relatively little changes by modifying from_flatbuf etc) or concatenating (concatenating would be ok since the problem only arises in few locations).

Having to postprocess flatbuffer files is awkward and the problem will probably come up again once we support more languages.

**Comment by jovany-wang** [comment]
Since this PR is outdated too much, I will request another one later and this could be closed now.

**Pull request closed without merging**